<?xml version="1.0" encoding="UTF-8"?>
<!--This does not replace URDF, and is not an extension of URDF.
    This is a format for representing semantic information about the robot structure.
    A URDF file must exist for this robot as well, where the joints and the links that are referenced are defined
-->
<robot name="Spot">
    <!--GROUPS: Representation of a set of joints and links. This can be useful for specifying DOF to plan for, defining arms, end effectors, etc-->
    <!--LINKS: When a link is specified, the parent joint of that link (if it exists) is automatically included-->
    <!--JOINTS: When a joint is specified, the child link of that joint (which will always exist) is automatically included-->
    <!--CHAINS: When a chain is specified, all the links along the chain (including endpoints) are included in the group. Additionally, all the joints that are parents to included links are also included. This means that joints along the chain and the parent joint of the base link are included in the group-->
    <!--SUBGROUPS: Groups can also be formed by referencing to already defined group names-->
    <group name="spot_arm">
        <joint name="gripper_centre_joint"/>
        <chain base_link="SpotArm" tip_link="front"/>
    </group>
    <group name="tiago_gripper">
        <joint name="gripper_left_finger_joint"/>
        <joint name="gripper_right_finger_joint"/>
    </group>
    <!--GROUP STATES: Purpose: Define a named state for a particular group, in terms of joint values. This is useful to define states like 'folded arms'-->
    <group_state name="retracted" group="spot_arm">
        <joint name="Slider11" value="0"/>
        <joint name="spotarm_1_joint" value="0"/>
        <joint name="spotarm_2_joint" value="3.1415"/>
        <joint name="spotarm_3_joint" value="3"/>
        <joint name="spotarm_4_joint" value="0"/>
        <joint name="spotarm_5_joint" value="0.2"/>
        <joint name="spotarm_6_joint" value="0"/>
    </group_state>
    <group_state name="open" group="tiago_gripper">
        <joint name="gripper_left_finger_joint" value="0.045"/>
        <joint name="gripper_right_finger_joint" value="0.045"/>
    </group_state>
    <group_state name="close" group="tiago_gripper">
        <joint name="gripper_left_finger_joint" value="0"/>
        <joint name="gripper_right_finger_joint" value="0"/>
    </group_state>
    <!--VIRTUAL JOINT: Purpose: this element defines a virtual joint between a robot link and an external frame of reference (considered fixed with respect to the robot)-->
    <virtual_joint name="gc" type="fixed" parent_frame="front" child_link="gripper_centre"/>
    <!--PASSIVE JOINT: Purpose: this element is used to mark joints that are not actuated-->
    <passive_joint name="front left shoulder abduction motor"/>
    <passive_joint name="front left shoulder rotation motor"/>
    <passive_joint name="front left elbow motor"/>
    <passive_joint name="front left piston motor"/>
    <passive_joint name="front right shoulder abduction motor"/>
    <passive_joint name="front right shoulder rotation motor"/>
    <passive_joint name="front right elbow motor"/>
    <passive_joint name="front right piston motor"/>
    <passive_joint name="rear left shoulder abduction motor"/>
    <passive_joint name="rear left shoulder rotation motor"/>
    <passive_joint name="rear left elbow motor"/>
    <passive_joint name="rear left piston motor"/>
    <passive_joint name="rear right shoulder abduction motor"/>
    <passive_joint name="rear right shoulder rotation motor"/>
    <passive_joint name="rear right elbow motor"/>
    <passive_joint name="rear right piston motor"/>
    <!--DISABLE COLLISIONS: By default it is assumed that any link of the robot could potentially come into collision with any other link in the robot. This tag disables collision checking between a specified pair of links. -->
    <disable_collisions link1="SpotArm" link2="Velodyne_Puck" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="base_link" reason="Adjacent"/>
    <disable_collisions link1="SpotArm" link2="front left forearm" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="front left shoulder" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="front left upperarm" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="front right forearm" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="front right shoulder" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="front right upperarm" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="gripper_left_finger_link" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="gripper_right_finger_link" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="rear left forearm" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="rear left shoulder" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="rear left upperarm" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="rear right forearm" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="rear right upperarm" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="spotarm_link1" reason="Adjacent"/>
    <disable_collisions link1="SpotArm" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="SpotArm" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="base_link" reason="Adjacent"/>
    <disable_collisions link1="Velodyne_Puck" link2="front left shoulder" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="front left upperarm" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="front right forearm" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="front right shoulder" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="front right upperarm" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="rear left shoulder" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="rear left upperarm" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="rear right forearm" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="rear right upperarm" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="Velodyne_Puck" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="base_link" link2="front left shoulder" reason="Adjacent"/>
    <disable_collisions link1="base_link" link2="front right shoulder" reason="Adjacent"/>
    <disable_collisions link1="base_link" link2="rear left shoulder" reason="Adjacent"/>
    <disable_collisions link1="base_link" link2="rear right shoulder" reason="Adjacent"/>
    <disable_collisions link1="base_link" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="front" link2="gripper_left_finger_link" reason="Adjacent"/>
    <disable_collisions link1="front" link2="gripper_right_finger_link" reason="Adjacent"/>
    <disable_collisions link1="front" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="front" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="front" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="front" link2="spotarm_link5" reason="Never"/>
    <disable_collisions link1="front" link2="spotarm_link6" reason="Adjacent"/>
    <disable_collisions link1="front left forearm" link2="front left shoulder" reason="Never"/>
    <disable_collisions link1="front left forearm" link2="front left upperarm" reason="Adjacent"/>
    <disable_collisions link1="front left forearm" link2="front right shoulder" reason="Never"/>
    <disable_collisions link1="front left forearm" link2="rear left shoulder" reason="Never"/>
    <disable_collisions link1="front left forearm" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="front left forearm" link2="rear right upperarm" reason="Never"/>
    <disable_collisions link1="front left forearm" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="front left upperarm" reason="Adjacent"/>
    <disable_collisions link1="front left shoulder" link2="front right forearm" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="front right upperarm" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="rear left forearm" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="rear left shoulder" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="rear left upperarm" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="rear right forearm" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="rear right upperarm" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="front left shoulder" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="front left upperarm" link2="front right shoulder" reason="Never"/>
    <disable_collisions link1="front left upperarm" link2="rear left shoulder" reason="Never"/>
    <disable_collisions link1="front left upperarm" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="front left upperarm" link2="rear right upperarm" reason="Never"/>
    <disable_collisions link1="front left upperarm" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="front right forearm" link2="front right shoulder" reason="Never"/>
    <disable_collisions link1="front right forearm" link2="front right upperarm" reason="Adjacent"/>
    <disable_collisions link1="front right forearm" link2="rear left shoulder" reason="Never"/>
    <disable_collisions link1="front right forearm" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="front right forearm" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="front right shoulder" link2="front right upperarm" reason="Adjacent"/>
    <disable_collisions link1="front right shoulder" link2="rear left forearm" reason="Never"/>
    <disable_collisions link1="front right shoulder" link2="rear left shoulder" reason="Never"/>
    <disable_collisions link1="front right shoulder" link2="rear left upperarm" reason="Never"/>
    <disable_collisions link1="front right shoulder" link2="rear right forearm" reason="Never"/>
    <disable_collisions link1="front right shoulder" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="front right shoulder" link2="rear right upperarm" reason="Never"/>
    <disable_collisions link1="front right shoulder" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="front right shoulder" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="front right shoulder" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="front right upperarm" link2="rear left shoulder" reason="Never"/>
    <disable_collisions link1="front right upperarm" link2="rear left upperarm" reason="Never"/>
    <disable_collisions link1="front right upperarm" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="front right upperarm" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="gripper_left_finger_link" link2="gripper_right_finger_link" reason="Never"/>
    <disable_collisions link1="gripper_left_finger_link" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="gripper_left_finger_link" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="gripper_left_finger_link" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="gripper_left_finger_link" link2="spotarm_link5" reason="Never"/>
    <disable_collisions link1="gripper_left_finger_link" link2="spotarm_link6" reason="Never"/>
    <disable_collisions link1="gripper_right_finger_link" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="gripper_right_finger_link" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="gripper_right_finger_link" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="gripper_right_finger_link" link2="spotarm_link5" reason="Never"/>
    <disable_collisions link1="gripper_right_finger_link" link2="spotarm_link6" reason="Never"/>
    <disable_collisions link1="rear left forearm" link2="rear left shoulder" reason="Never"/>
    <disable_collisions link1="rear left forearm" link2="rear left upperarm" reason="Adjacent"/>
    <disable_collisions link1="rear left forearm" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="rear left forearm" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="rear left shoulder" link2="rear left upperarm" reason="Adjacent"/>
    <disable_collisions link1="rear left shoulder" link2="rear right forearm" reason="Never"/>
    <disable_collisions link1="rear left shoulder" link2="rear right upperarm" reason="Never"/>
    <disable_collisions link1="rear left shoulder" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="rear left shoulder" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="rear left shoulder" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="rear left upperarm" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="rear left upperarm" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="rear left upperarm" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="rear left upperarm" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="rear right forearm" link2="rear right shoulder" reason="Never"/>
    <disable_collisions link1="rear right forearm" link2="rear right upperarm" reason="Adjacent"/>
    <disable_collisions link1="rear right forearm" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="rear right forearm" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="rear right shoulder" link2="rear right upperarm" reason="Adjacent"/>
    <disable_collisions link1="rear right shoulder" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="rear right shoulder" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="rear right shoulder" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="rear right upperarm" link2="spotarm_link1" reason="Never"/>
    <disable_collisions link1="rear right upperarm" link2="spotarm_link2" reason="Never"/>
    <disable_collisions link1="rear right upperarm" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="rear right upperarm" link2="spotarm_link5" reason="Never"/>
    <disable_collisions link1="spotarm_link1" link2="spotarm_link2" reason="Adjacent"/>
    <disable_collisions link1="spotarm_link1" link2="spotarm_link3" reason="Never"/>
    <disable_collisions link1="spotarm_link1" link2="spotarm_link5" reason="Never"/>
    <disable_collisions link1="spotarm_link1" link2="spotarm_link6" reason="Never"/>
    <disable_collisions link1="spotarm_link2" link2="spotarm_link3" reason="Adjacent"/>
    <disable_collisions link1="spotarm_link2" link2="spotarm_link5" reason="Never"/>
    <disable_collisions link1="spotarm_link2" link2="spotarm_link6" reason="Never"/>
    <disable_collisions link1="spotarm_link3" link2="spotarm_link5" reason="Adjacent"/>
    <disable_collisions link1="spotarm_link3" link2="spotarm_link6" reason="Never"/>
    <disable_collisions link1="spotarm_link5" link2="spotarm_link6" reason="Adjacent"/>
</robot>
